File: [H]array/[H]array-prefix-sum/560-subarray-sum-equals-k.py
from collections import defaultdict
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        prefix_count = defaultdict(int)
        prefix_count[0] = 1
        total = 0
        res = 0
        for num in nums:
            total += num
            res += prefix_count[total - k]
            prefix_count[total] += 1
        return res
    
# time O(n)
# space O(n)
# using array and prefix sum and hashmap to validate the gap subarray
'''
1. if meet the prefix - k in hashmap means the gap is valid
'''

File: [H]array/[H]array-prefix-sum/560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int res = 0;
        HashMap<Integer, Integer> valToFreq = new HashMap<>();
        valToFreq.put(0, 1);
        int total = 0;
        for (int num: nums) {
            total += num;
            if (valToFreq.containsKey(total - k)) {
                res += valToFreq.get(total - k);
            }
            valToFreq.compute(total, (key, v) -> v == null ? 1 : v + 1);
        }
        return res;
    }
}

// time O(n)
// space O(n)
// using array and prefix sum and hashmap to validate the gap subarray
/*
1. if meet the prefix - k in hashmap means the gap is valid
*/